### PR TITLE
Introduce `extend` and `from`

### PR DESCRIPTION
**Pull request by wwilsman**
## Purpose

Now that [decorators have landed in Stage 2](https://babeljs.io/blog/2018/09/17/decorators), the decorator function arguments have changed from legacy decorators in Stage 1.

The argument for legacy class decorators was the target constructor. Returning a new constructor replaced the decorated constructor completely and this was how interactor classes were created when using the interaction creators.

The new argument for Stage 2 decorators is a class element descriptor, which provides access to all own initializers without having to construct an instance of the original class. Element descriptors are plain objects with specific properties. To replace the constructor, you must return a class element descriptor with a `finisher` function. For more info, [check out the proposal](https://github.com/tc39/proposal-decorators/blob/master/METAPROGRAMMING.md#class-decorator).

Now that the class decorator argument is a plain object, that presents potential issues when differentiating between an element descriptor and an object consisting of interactor properties. So we should split this functionality and have a separate method that handles plain objects, instead of overloading the decorator.

## Approach

I already had plans to figure out a way to make the decorator extensible so that custom interactors could inherit from other custom interactors. This seemed like the perfect time to do so.

First, the plain object syntax was pulled out into its own static method. Since extending custom interactors would use the `extend` keyword, and we are avoiding overloading the decorator, I opted for `from` to convert plain objects into interactor classes.

Second, the decorator was moved to a static `extend` method and now the existing `interactor` decorator calls on `Interactor.extend`. The original logic for the decorator changed as well. Before, it collected properties from an instance (or an object) and defined them on the returned custom interactor. Now, the properties are collected from the class element descriptors and passed along to the `from` method.

The legacy decorator is still supported by checking if the class descriptor is actually a constructor. It then has its own properties and descriptors copied to a plain object before passing it along to `from` as well. Using `extend` on plain objects still works, but is deprecated since it is meant to be a pure decorator.

While working on this, I refactored the `utils` file after moving the parent-chainable logic into its own util file. This revealed that many of the utils were actually only used by other utils or the decorator. I also cleaned up some dependencies and removed a `karma-webpack` issue by replacing Karma with the BigTest CLI.

### Future

I chose not to deprecate the legacy decorator syntax right now to give users the chance to migrate over from using it on plain objects first. In the next version, we should deprecate the legacy syntax in favor of the Stage 2 syntax. And for 1.0 the decorator will only work according to the current proposal.

### CODE REVIEW

**Comment by Robdel12** [comment]
> removed a karma-webpack issue by replacing Karma with the BigTest CLI.

❤️

File: src/utils/extend.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+              name: constructor.name
+            }, constructor)
+          })
+        );
+      }
+    };
+
+  // a class constructor was provided (legacy decorator syntax)
+  } else if (typeof classDescriptor === 'function') {
+    // console.warn(`Deprecated. Please upgrade to Stage 2 decorators.`);
```
[the remark this answers is not in the document]
Comment (wwilsman) [reply src/utils/extend.js]: > I chose not to deprecate the legacy decorator syntax right now [until the] next version

😉

File: src/utils/extend.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+ * @param {String} key
+ * @returns {Object}
+ */
+function omit(obj, key) {
+  let { [key]: _, ...rest } = obj;
+  return rest;
+}
+
+/**
+ * Similar to the `\@interactor` decorator; creates a custom interactor class
```
[the remark this answers is not in the document]
Comment (wwilsman) [reply src/utils/extend.js]: Better safe than sorry

File: src/utils/extend.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Robdel12) [reply src/utils/extend.js]: Ahh, missed that part

**Pull request merged**